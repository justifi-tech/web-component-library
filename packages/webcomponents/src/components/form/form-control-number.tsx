import {
  Component,
  Host,
  h,
  Prop,
  Event,
  EventEmitter,
  State,
  Element,
  Watch,
} from '@stencil/core';

@Component({
  tag: 'form-control-number'
})
export class NumberInput {
  @Element() el: HTMLElement;

  @Prop() label: string;
  @Prop() name: any;
  @Prop() helpText?: string;
  @Prop() errorText?: string;
  @Prop() defaultValue: string;
  @Prop() inputHandler: (name: string, value: string) => void;
  @Prop() disabled: boolean;

  @State() input: string;
  @Event() formControlInput: EventEmitter<any>;
  @Event() formControlBlur: EventEmitter<any>;

  @Watch('defaultValue')
  handleDefaultValueChange(newValue: string) {
    this.updateInput(newValue);
  }

  updateInput(newValue: any) {
    const inputElement = this.el.querySelector('input');
    if (inputElement) {
      inputElement.value = newValue || '';
    }
  }

  handleFormControlInput(event: any) {
    const target = event.target;
    const name = target.getAttribute('name');
    this.inputHandler(name, target.value);
    this.formControlInput.emit({ name, value: target.value });
  }

  componentDidLoad() {
    this.updateInput(this.defaultValue);
  }

  render() {
    return (
      <Host exportparts='label,input,input-invalid'>
        <div class='form-group d-flex flex-column'>
          <label part="label" class="form-label" htmlFor={this.name}>
            {this.label}
          </label>
          <input
            id={this.name}
            name={this.name}
            onInput={(event: any) => this.handleFormControlInput(event)}
            onBlur={() => this.formControlBlur.emit()}
            part={`input ${this.errorText && 'input-invalid'}`}
            class={this.errorText ? 'form-control is-invalid' : 'form-control'}
            type="number"
            disabled={this.disabled}
          />
          <form-control-help-text helpText={this.helpText} name={this.name} />
          <form-control-error-text errorText={this.errorText} name={this.name} />
        </div>
      </Host>
    );
  }
}
