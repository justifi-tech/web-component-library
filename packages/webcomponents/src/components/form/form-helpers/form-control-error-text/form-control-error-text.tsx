import { Component, Prop, h } from '@stencil/core';

@Component({
  tag: 'form-control-error-text',
})
export class FormControlErrorText {
  @Prop() errorText: string;
  @Prop() name: string;

  render() {

    if (!this.errorText) {
      return null;
    }

    return (
      <small class='form-text text-danger' id={`form-error-text-${this.name}`}>
        {this.errorText}
      </small>
    );
  }
}
