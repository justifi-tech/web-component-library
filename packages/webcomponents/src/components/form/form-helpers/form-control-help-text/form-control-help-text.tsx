import { Component, Prop, h } from '@stencil/core';

@Component({
  tag: 'form-control-help-text',
})
export class FormControlHelpText {
  @Prop() helpText: string;
  @Prop() errorText: string;

  private text: string;

  get colorClass() {
    return this.errorText ? 'text-danger' : 'text-muted';
  }

  render() {

    if (!this.helpText && !this.errorText) {
      return null;
    }

    if (this.errorText) {
      this.text = this.errorText;
    } else {
      this.text = this.helpText;
    }

    return (
      <small class={`form-text ${this.colorClass}`} id='formHelpText'>
        {this.text}
      </small>
    );
  }
}
