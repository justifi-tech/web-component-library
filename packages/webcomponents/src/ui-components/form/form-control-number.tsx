import {
  Component,
  Host,
  h,
  Prop,
  Event,
  EventEmitter,
  Element,
  Watch,
} from '@stencil/core';
import { FormControlErrorText, FormControlHelpText } from '../../ui-components';

@Component({
  tag: 'form-control-number'
})
export class NumberInput {
  @Element() el: HTMLElement;

  @Prop() label: string;
  @Prop() name: any;
  @Prop() helpText?: string;
  @Prop() errorText?: string;
  @Prop() defaultValue: string;
  @Prop() inputHandler: (name: string, value: string) => void;
  @Prop() disabled: boolean;

  @Watch('defaultValue')
  handleDefaultValueChange(newValue: string) {
    this.updateInput(newValue);
  }

  @Event() formControlInput: EventEmitter<any>;
  @Event() formControlBlur: EventEmitter<any>;

  componentDidLoad() {
    this.updateInput(this.defaultValue);
  }

  handleFormControlInput = (event: any) => {
    const target = event.target;
    const name = target.getAttribute('name');
    this.inputHandler(name, target.value);
    this.formControlInput.emit({ name, value: target.value });
  }

  updateInput = (newValue: any) => {
    const inputElement = this.el.querySelector('input');
    if (inputElement) {
      inputElement.value = newValue || '';
    }
  }

  render() {
    return (
      <Host exportparts='label,input,input-invalid'>
        <div class='form-group d-flex flex-column'>
          <label part="label" class="form-label" htmlFor={this.name}>
            {this.label}
          </label>
          <input
            id={this.name}
            name={this.name}
            onBlur={this.formControlBlur.emit}
            onInput={this.handleFormControlInput}
            part={`input ${this.errorText && 'input-invalid'}`}
            class={this.errorText ? 'form-control is-invalid' : 'form-control'}
            type="number"
            disabled={this.disabled}
          />
          <FormControlHelpText helpText={this.helpText} name={this.name} />
          <FormControlErrorText errorText={this.errorText} name={this.name} />
        </div>
      </Host>
    );
  }
}
