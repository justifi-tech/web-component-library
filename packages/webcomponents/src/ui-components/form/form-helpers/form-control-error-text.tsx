import { h } from '@stencil/core';

export const FormControlErrorText = (props: { errorText: string, name: string }) => {
  const { errorText, name } = props;

  if (!errorText) {
    return null;
  }

  return (
    <small class='form-text text-danger' id={`form-error-text-${name}`} part="input-error-text" >
      {errorText}
    </small>
  );
}
