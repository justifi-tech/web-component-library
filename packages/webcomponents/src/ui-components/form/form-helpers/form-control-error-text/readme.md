# form-control-error-text



<!-- Auto Generated Below -->


## Properties

| Property    | Attribute    | Description | Type     | Default     |
| ----------- | ------------ | ----------- | -------- | ----------- |
| `errorText` | `error-text` |             | `string` | `undefined` |
| `name`      | `name`       |             | `string` | `undefined` |


## Dependencies

### Used by

 - [form-control-checkbox](../..)
 - [form-control-date](../..)
 - [form-control-file](../..)
 - [form-control-file-v2](../..)
 - [form-control-monetary](../..)
 - [form-control-number](../..)
 - [form-control-number-masked](../..)
 - [form-control-select](../..)
 - [form-control-text](../..)
 - [form-control-textarea](../..)
 - [justifi-dispute-reason](../../../../components/dispute-management/dispute-response)

### Graph
```mermaid
graph TD;
  form-control-checkbox --> form-control-error-text
  form-control-date --> form-control-error-text
  form-control-file --> form-control-error-text
  form-control-file-v2 --> form-control-error-text
  form-control-monetary --> form-control-error-text
  form-control-number --> form-control-error-text
  form-control-number-masked --> form-control-error-text
  form-control-select --> form-control-error-text
  form-control-text --> form-control-error-text
  form-control-textarea --> form-control-error-text
  justifi-dispute-reason --> form-control-error-text
  style form-control-error-text fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
