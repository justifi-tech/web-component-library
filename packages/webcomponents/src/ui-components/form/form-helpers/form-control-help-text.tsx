import { h } from '@stencil/core';

export const FormControlHelpText = (props: { helpText: string, name: string }) => {
  const { helpText, name } = props;

  if (!helpText) {
    return null;
  }

  return (
    <small class='form-text text-muted' id={`form-help-text-${name}`} part="input-help-text">
      {helpText}
    </small>
  );
}
