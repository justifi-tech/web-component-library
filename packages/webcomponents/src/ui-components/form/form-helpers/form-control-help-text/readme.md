# form-control-help-text



<!-- Auto Generated Below -->


## Properties

| Property   | Attribute   | Description | Type     | Default     |
| ---------- | ----------- | ----------- | -------- | ----------- |
| `helpText` | `help-text` |             | `string` | `undefined` |
| `name`     | `name`      |             | `string` | `undefined` |


## Dependencies

### Used by

 - [bank-account-form-inputs](../../../../components/business-forms/payment-provisioning/bank-account)
 - [bank-account-form-inputs-canada](../../../../components/business-forms/payment-provisioning/bank-account)
 - [form-control-radio](../..)
 - [form-control-textarea](../..)

### Graph
```mermaid
graph TD;
  bank-account-form-inputs --> form-control-help-text
  bank-account-form-inputs-canada --> form-control-help-text
  form-control-radio --> form-control-help-text
  form-control-textarea --> form-control-help-text
  style form-control-help-text fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
