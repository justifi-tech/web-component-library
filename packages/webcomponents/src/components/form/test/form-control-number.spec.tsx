import { h } from '@stencil/core';
import { newSpecPage } from "@stencil/core/testing";
import { NumberInput } from "../form-control-number";
import { FormControlErrorText } from '../form-helpers/form-control-error-text/form-control-error-text';
import { FormControlHelpText } from '../form-helpers/form-control-help-text/form-control-help-text';

describe('form-control-number', () => {
  const components = [NumberInput, FormControlErrorText, FormControlHelpText];
  const mockInputHandler = jest.fn();

  it('Renders with default props', async () => {
    const page = await newSpecPage({
      components: components,
      template: () => <form-control-number label='Age' name='age'></form-control-number>,
    });

    expect(page.root).toMatchSnapshot();
  });

  it('Renders with all props provided', async () => {
    const page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-number
          label='Age'
          name='age'
          defaultValue='25'
          errorText='Invalid age'
          helpText='Enter your age'
          disabled
          inputHandler={mockInputHandler}
        >
        </form-control-number>
    });

    expect(page.root).toMatchSnapshot();
  });

  it('Updates the input value when defaultValue changes', async () => {
    // Initial render with the default value
    let defaultValue = '25';
    let page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-number
          label='Age'
          name='age'
          defaultValue={defaultValue}
          inputHandler={mockInputHandler}
        >
        </form-control-number>
    });
    let inputElement = page.root.querySelector('input');
    expect(inputElement.value).toBe('25');

    defaultValue = '30';

    page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-number
          label='Age'
          name='age'
          defaultValue={defaultValue}
          inputHandler={mockInputHandler}
        >
        </form-control-number>
    });

    inputElement = page.root.querySelector('input');
    expect(inputElement.value).toBe('30');
  });

  it('Handles user input correctly', async () => {
    const page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-number
          label='Age'
          name='age'
          defaultValue='25'
          inputHandler={mockInputHandler}
        >
        </form-control-number>
    });

    const inputElement = page.root.querySelector('input');
    const testValue = '30';

    inputElement.value = testValue;
    await inputElement.dispatchEvent(new Event('input'));

    expect(inputElement.value).toBe(testValue);
  });

  it('Emits formControlInput event on input', async () => {
    const page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-number
          label='Age'
          name='age'
          defaultValue='25'
          inputHandler={mockInputHandler}
        >
        </form-control-number>
    });

    const inputEventSpy = jest.fn();
    page.win.addEventListener('formControlInput', inputEventSpy);

    const inputElement = page.root.querySelector('input');
    inputElement.value = '25';

    inputElement.dispatchEvent(new Event('input', { bubbles: true, composed: true }));

    expect(inputEventSpy).toHaveBeenCalled();
  });

  it('Emits formControlBlur event on blur', async () => {
    const page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-number
          label='Age'
          name='age'
          inputHandler={mockInputHandler}
        >
        </form-control-number>
    });
    const blurEventSpy = jest.fn();
    page.win.addEventListener('formControlBlur', blurEventSpy);

    const inputElement = page.root.querySelector('input');
    await inputElement.dispatchEvent(new CustomEvent('blur'));

    expect(blurEventSpy).toHaveBeenCalled();
  });

  it('Disables input when disabled prop is provided', async () => {
    const page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-number
          label='Age'
          name='age'
          disabled
          inputHandler={mockInputHandler}
        >
        </form-control-number>
    });

    const inputElement = page.root.querySelector('input');
    expect(inputElement.disabled).toBe(true);
  });

  it('Shows help text when helpText prop is provided', async () => {
    const page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-number
          label='Age'
          name='age'
          helpText='Enter your age'
          inputHandler={mockInputHandler}
        >
        </form-control-number>
    });

    const helpTextComponent = page.root.querySelector('form-control-help-text');
    expect(helpTextComponent).not.toBeNull();

    const helpText = helpTextComponent.querySelector('.text-muted');
    expect(helpText.textContent).toBe('Enter your age');
  });

  it('Shows error and applies error styling when error prop is provided', async () => {
    const page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-number
          label='Age'
          name='age'
          errorText='Invalid age'
          inputHandler={mockInputHandler}
        >
        </form-control-number>
    });

    const errorTextComponent = page.root.querySelector('form-control-error-text');
    expect(errorTextComponent).not.toBeNull();

    const errorText = errorTextComponent.querySelector('.text-danger');
    expect(errorText.textContent).toBe('Invalid age');

    const inputElement = page.root.querySelector('input');
    expect(inputElement.classList.contains('is-invalid')).toBe(true);
  });
});
