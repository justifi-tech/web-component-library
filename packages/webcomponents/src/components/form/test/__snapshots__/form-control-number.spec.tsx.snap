// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form-control-number handles the disabled state 1`] = `
<form-control-number disabled="" exportparts="label,input,input-invalid" label="Age">
  <mock:shadow-root>
    <label class="form-label" part="label">
      Age
    </label>
    <input class="form-control" disabled="" part="input undefined" type="number" value="">
  </mock:shadow-root>
</form-control-number>
`;

exports[`form-control-number renders with an error message 1`] = `
<form-control-number error="Invalid input" exportparts="label,input,input-invalid" label="Age">
  <mock:shadow-root>
    <label class="form-label" part="label">
      Age
    </label>
    <input class="form-control is-invalid" part="input input-invalid" type="number" value="">
    <div class="invalid-feedback">
      Invalid input
    </div>
  </mock:shadow-root>
</form-control-number>
`;

exports[`form-control-number renders with default props 1`] = `
<form-control-number exportparts="label,input,input-invalid" label="Age">
  <mock:shadow-root>
    <label class="form-label" part="label">
      Age
    </label>
    <input class="form-control" part="input undefined" type="number" value="">
  </mock:shadow-root>
</form-control-number>
`;
